pkg/kube/secretcontroller/secretcontroller.go

```diff
@@ -0,0 +1,269 @@
+	log.Info("Waiting for informer caches to sync")
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+	wait.Until(c.runWorker, 5*time.Second, stopCh)
+}
+
+// StartSecretController creates the secret controller.
+func StartSecretController(k8s kubernetes.Interface,
+	addfn AddClusterFunc,
+	rmfn RmClusterFunc,
+	namespace string) error {
+	stopCh := make(chan struct{})
+	clusterStore := newClustersStore()
+	controller := NewController(k8s, namespace, clusterStore, addfn, rmfn)
+
+	go controller.Run(stopCh)
+
+	return nil
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextItem() {
+		// continue looping
+	}
+}
+
+func (c *Controller) processNextItem() bool {
+	secretName, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(secretName)
+
+	err := c.processItem(secretName.(string))
+	if err == nil {
+		// No error, reset the ratelimit counters
+		c.queue.Forget(secretName)
+	} else if c.queue.NumRequeues(secretName) < maxRetries {
+		log.Errorf("Error processing %s (will retry): %v", secretName, err)
+		c.queue.AddRateLimited(secretName)
+	} else {
+		log.Errorf("Error processing %s (giving up): %v", secretName, err)
+		c.queue.Forget(secretName)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+func (c *Controller) processItem(secretName string) error {
+	obj, exists, err := c.informer.GetIndexer().GetByKey(secretName)
+	if err != nil {
+		return fmt.Errorf("Error fetching object %s error: %v", secretName, err)
+	}
+
+	if exists {
+		c.addMemberCluster(secretName, obj.(*corev1.Secret))
+	} else {
+		c.deleteMemberCluster(secretName)
+	}
+
+	return nil
+}
+
+func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
+	c.cs.StoreLock.Lock()
+	defer c.cs.StoreLock.Unlock()
+
+	for clusterID, kubeConfig := range s.Data {
+		// custerID must be unique even across multiple secrets
+		if _, ok := c.cs.Rc[clusterID]; !ok {
+
+			if len(kubeConfig) == 0 {
+				log.Infof("Data '%s' in the secret %s in namespace %s is empty, and disregarded ",
+					clusterID, secretName, s.ObjectMeta.Namespace)
+				continue
+			}
+
+			clientConfig, err := loadKubeConfig(kubeConfig)
+
+			if err != nil {
+				log.Infof("Data '%s' in the secret %s in namespace %s is not a kubeconfig: %v",
+					clusterID, secretName, s.ObjectMeta.Namespace, err)
+				continue
+			}
+
+			log.Infof("Adding new cluster member: %s", clusterID)
+			c.cs.Rc[clusterID] = &RemoteCluster{}
+			c.cs.Rc[clusterID].FromSecret = secretName
+			client, _ := kube.CreateInterfaceFromClusterConfig(clientConfig)
+			err = c.addcallback(client, clusterID)
+			if err != nil {
+				log.Errorf("Error during create of clusterID: %s %v", clusterID, err)
+			}
+
+		} else {
+			log.Infof("Cluster %s in the secret %s in namespace %s already exists",
+				clusterID, secretName, s.ObjectMeta.Namespace)
```

Comment:
secretName should be the one from c.cs.Rc[clusterID].

Reply:
done